File: Medium/Job Sequencing Problem/job-sequencing-problem.cpp
//{ Driver Code Starts
// Program to find the maximum profit job sequence from a given array 
// of jobs with deadlines and profits 
#include<bits/stdc++.h>
using namespace std; 

// A structure to represent a job 
struct Job 
{ 
    int id;	 // Job Id 
    int dead; // Deadline of job 
    int profit; // Profit if job is over before or on deadline 
}; 


// } Driver Code Ends
/*
struct Job 
{ 
    int id;	 // Job Id 
    int dead; // Deadline of job 
    int profit; // Profit if job is over before or on deadline 
};
*/
bool comparator(Job arr1, Job arr2)
{
    return (arr1.profit > arr2.profit);
    
    
}

class Solution 
{
    public:
    //Function to find the maximum profit and the number of jobs done.
    vector<int> JobScheduling(Job arr[], int n) 
    { 
        sort(arr,arr+n,comparator);
        
        int max_profit=0,max_deadline=arr[0].dead,jobs=0;
        
        for(int i=1;i<n;i++)
        {
            max_deadline = max(max_deadline,arr[i].dead);
        }
        
        vector<int> profit(max_deadline+1,-1);
        
        for(int i=0;i<n;i++)
        {
            
            for(int j=arr[i].dead;j>0;j--)
            {
                if(profit[j]==-1)
                {
                    max_profit+=arr[i].profit;
                    jobs++;
                    profit[j]=i;
                    break;
                }
            }
        }
        return {jobs,max_profit};
        // your code here
    } 
};

//{ Driver Code Starts.
// Driver program to test methods 
int main() 
{ 
    int t;
    //testcases
    cin >> t;
    
    while(t--){
        int n;
        
        //size of array
        cin >> n;
        Job arr[n];
        
        //adding id, deadline, profit
        for(int i = 0;i<n;i++){
                int x, y, z;
                cin >> x >> y >> z;
                arr[i].id = x;
                arr[i].dead = y;
                arr[i].profit = z;
        }
        Solution ob;
        //function call
        vector<int> ans = ob.JobScheduling(arr, n);
        cout<<ans[0]<<" "<<ans[1]<<endl;
    }
	return 0; 
}



// } Driver Code Ends

File: Difficulty: Medium/Job Sequencing Problem/job-sequencing-problem.py
#User function Template for python3
'''
class Job:
    
    # Job class which stores profit and deadline.
    
    def __init__(self,profit=0,deadline=0):
        self.profit = profit
        self.deadline = deadline
        self.id = 0
'''        

class Solution:
    
    #Function to find the maximum profit and the number of jobs done.
    def JobScheduling(self,Jobs,n):
        Jobs.sort(key = lambda x: x.profit,reverse=True)
        max_deadline = max([x.deadline for x in Jobs])
        
        # for x in Jobs:
        #     print(x.id,x.profit,x.deadline)
        jobq = [-1]*(max_deadline+1)
        
        profit = 0
        count = 0
        for x in Jobs:
            if jobq[x.deadline] == -1:
                jobq[x.deadline] = x.id
                profit += x.profit
                count+=1
            else:
                i=x.deadline-1
                
                
                while i > 0:
                    if jobq[i] == -1:
                        # print(x.profit,x.id)
                        jobq[i] = x.id
                        profit += x.profit
                        count+=1
                        break
                    i-=1
                    
                    
        # print(jobq)
        
        return [count,profit]
        
        # code here


#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys


#Contributed by : Nagendra Jha
class Job:
    '''
    Job class which stores profit and deadline.
    '''

    def __init__(self, profit=0, deadline=0):
        self.profit = profit
        self.deadline = deadline
        self.id = 0


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())
        info = list(map(int, input().strip().split()))
        Jobs = [Job() for i in range(n)]
        for i in range(n):
            Jobs[i].id = info[3 * i]
            Jobs[i].deadline = info[3 * i + 1]
            Jobs[i].profit = info[3 * i + 2]
        ob = Solution()
        res = ob.JobScheduling(Jobs, n)
        print(res[0], end=" ")
        print(res[1])

# } Driver Code Ends